File: S/922-sort-array-by-parity-ii/main.go
package main

import "fmt"

func sortArrayByParityII(A []int) []int {

	var ODD, EVEN []int
	for _, v := range A {
		if v%2 == 0 {
			EVEN = append(EVEN, v)
		} else {
			ODD = append(ODD, v)
		}
	}
	sO, sE := 0, 0
	for i := 0; i < len(A); i++ {
		if i%2 == 0 {
			A[i] = EVEN[sE]
			sE++
		} else {
			A[i] = ODD[sO]
			sO++
		}
	}
	return A
}

func main() {
	fmt.Println("a")
}

File: S/922-sort-array-by-parity-ii/main.rs
struct Solution {}

impl Solution {
  pub fn sort_array_by_parity_ii(nums: Vec<i32>) -> Vec<i32> {
    let mut nums = nums;
    let mut even: usize = 0;
    let mut odd: usize = 1;

    while even < nums.len() && odd < nums.len() {
      while even < nums.len() && nums[even] % 2 == 0 {
        even += 2;
      }
      while odd < nums.len() && nums[odd] % 2 == 1 {
        odd += 2;
      }
      if even >= nums.len() || odd >= nums.len() {
        break;
      }
      let t = nums[even];
      nums[even] = nums[odd];
      nums[odd] = t;
      even += 2;
      odd += 2;
    }
    nums
  }
}
